common/lib/xmodule/xmodule/tests/__init__.py

```diff
@@ -250,24 +277,87 @@ def bulk_assertions(self):
+    def _shorten(self, message):
+        """
+        Limit a message to a particular length.
+        """
+        if len(message) > self.BULK_ASSERT_MSG_LENGTH:
+            return message[:self.BULK_ASSERT_MSG_LENGTH - 3] + '...'
+        else:
+            return message
+
+    @contextmanager
+    def _capture_assertion_errors(self):
+        try:
+            # Only wrap the first layer of assert functions by stashing away the manager
+            # before executing the assertion.
+            manager = self._manager
+            self._manager = None
+            yield
+        except AssertionError:  # pylint: disable=broad-except
+            if manager is not None:
+                # Reconstruct the stack in which the error was thrown (so that the traceback)
+                # isn't cut off at `assertion(*args, **kwargs)`.
+                type, value, tb = sys.exc_info()
+
+                # Count the number of stack frames before you get to a
+                # unittest context (walking up the stack from here).
+                relevant_frames = 0
+                for frame_record in inspect.stack():
+                    # This is the same criterion used by unittest to decide if a
+                    # stack frame is relevant to exception printing.
+                    frame = frame_record[0]
+                    if '__unittest' in frame.f_globals:
+                        break
+                    relevant_frames += 1
+
+                stack_above = traceback.extract_stack()[-relevant_frames:-1]
+
+                stack_below = traceback.extract_tb(tb)
+                formatted_stack = traceback.format_list(stack_above + stack_below)
+                formatted_exc = traceback.format_exception_only(type, value)
+                manager.log_error(
+                    "".join(formatted_stack + formatted_exc)
+                )
+            else:
+                raise
+        finally:
+            self._manager = manager
+
+    def _wrap_assertion(self, assertion):
+        @wraps(assertion)
+        def assert_(*args, **kwargs):
+            context = None
+
+            # Run the assertion, and capture any raised assertionErrors
+            with self._capture_assertion_errors():
+                context = assertion(*args, **kwargs)
+
+            # Handle the assertRaises family of functions by returning
+            # a context manager that surrounds the assertRaises
+            # with our assertion capturing context manager.
+            if context is not None:
+                return nested(self._capture_assertion_errors(), context)
+
+        return assert_
+
+    def __getattribute__(self, name):
+        """
+        Wrap all assert* methods of this class using self._wrap_assertion,
+        to capture all assertion errors in bulk.
+        """
+        base_attr = super(BulkAssertionTest, self).__getattribute__(name)
+        if name.startswith('assert'):
+            return self._wrap_assertion(base_attr)
```

Comment:
If we're  looking to pull things out of this PR, this change to bulk assertions seems separable.